### PR TITLE
Adapt bootloader to console; add custom TCL script for Vivado; Fix FPGA_GRABBED and renamed it to BOARD_GRABBED;

### PR DESCRIPTION
**Pull request by PedroAntunes178**
This PR can be accepted after PR https://github.com/IObundle/iob-soc/pull/513;

From that PR this one tackles:
- The $(FPGA_GRABBED) is set to `FPGA_GRABBED=/tmp/$(BOARD).grabbed`
- Related to issue https://github.com/IObundle/iob-lib/issues/413 the system runs when `INIT_MEM=0`, however currently when run with `INIT_MEM=0 EXT_MEM=1` the system blocks after the console sends the file. Therefor it is WIP.

### CODE REVIEW

File: iob_soc_setup.py [hunk 1]
```diff
@@ -51,11 +48,9 @@
 
 confs = \
 [
-    # SoC defines
-
     # SoC macros
     {'name':'INIT_MEM',      'type':'M', 'val':'1', 'min':'0', 'max':'1', 'descr':"Enable memory initialization"},
-#    {'name':'RUN_EXTMEM',    'type':'M', 'val':'0', 'min':'0', 'max':'1', 'descr':"Run firmware from external memory"}, # USE_DDR was merged with RUN_EXTMEM
+    {'name':'RUN_EXTMEM',    'type':'M', 'val':'NA', 'min':'0', 'max':'1', 'descr':"Run firmware from external memory"},
```
[the remark this answers is not in the document]
Comment (arturum1) [reply iob_soc_setup.py:53]: Another weird issue that seems related to this is that verilator and icarus checks are currently failing in the `make setup SETUP_ARGS="INIT_MEM=1"` test (as currently RUN_EXTMEM=1 by default it should also be running with external memory). However, the system is running without DDR, as the logs show:
```
diff test.log src/test.expected
157
8a9
158
> IOb-Bootloader: DDR in use and program runs from DDR
```
The test.log does not have the `IOb-Bootloader: DDR in use and program runs from DDR` message.

[the remark this answers is not in the document]
Comment (arturum1) [reply iob_soc_setup.py:53]: I tried setting the RUN_EXTMEM as default in my fork and running the command:  `make setup SETUP_ARGS="INIT_MEM=1"`
 It works corretly in my fork, printing the `IOb-Bootloader: DDR in use and program runs from DDR` message.

[the remark this answers is not in the document]
Comment (arturum1) [reply iob_soc_setup.py:53]: **UPDATE**: I have now seen the changes in https://github.com/IObundle/iob-lib/pull/417. The issue with the icarus and verilator checks is that the sim_setup.py and fpga_setup.py are no longer working correctly after the LIB update of that PR.
Those scripts need to be updated to check for the `NA` value of the macros.

File: hardware/simulation/sim_setup.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+# Delete all test*.expected files from build_dir
+dirpath=os.path.join(setup_module.build_dir, "hardware/simulation/src")
+for file in os.listdir(dirpath):
+    if file.startswith("test") and file.endswith(".expected"):
+        os.remove(os.path.join(dirpath,file))
+
+# Find out correct test.expected filename
+test_file_name='test'
+#Check if setup with INIT_MEM (check if macro exists)
+if next((i for i in setup_module.confs if i['name']=='INIT_MEM'), False):
```
Comment (arturum1) [inline hardware/simulation/sim_setup.py:15]:
```suggestion
if next((i for i in setup_module.confs if i['name']=='INIT_MEM'), False) and setup_module.confs['INIT_MEM']['val'] != 'NA':
```

File: hardware/simulation/sim_setup.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    test_file_name+='_initmem'
+#Check if setup with RUN_EXTMEM (check if macro exists)
+if next((i for i in setup_module.confs if i['name']=='RUN_EXTMEM'), False):
```
Comment (arturum1) [inline hardware/simulation/sim_setup.py:18]:
```suggestion
if next((i for i in setup_module.confs if i['name']=='RUN_EXTMEM'), False) and setup_module.confs['RUN_EXTMEM']['val'] != 'NA':
```

File: hardware/fpga/fpga_setup.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
+# This script is called during setup.
+# You can use 'setup_module' to access the contents of the iob_soc_setup.py python module
+import os
+import shutil
+
+# Find out correct test.expected filename
+test_file_name='test'
+#Check if setup with INIT_MEM (check if macro exists)
+if next((i for i in setup_module.confs if i['name']=='INIT_MEM'), False):
```
Comment (arturum1) [inline hardware/fpga/fpga_setup.py:9]:
```suggestion
if next((i for i in setup_module.confs if i['name']=='INIT_MEM'), False) and setup_module.confs['INIT_MEM']['val'] != 'NA':
```

File: hardware/fpga/fpga_setup.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    test_file_name+='_initmem'
+#Check if setup with RUN_EXTMEM (check if macro exists)
+if next((i for i in setup_module.confs if i['name']=='RUN_EXTMEM'), False):
```
Comment (arturum1) [inline hardware/fpga/fpga_setup.py:12]:
```suggestion
if next((i for i in setup_module.confs if i['name']=='RUN_EXTMEM'), False) and setup_module.confs['RUN_EXTMEM']['val'] != 'NA':
```

**Review by arturum1 — CHANGES_REQUESTED** [review]

File: hardware/simulation/sim_setup.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+# Delete all test*.expected files from build_dir
+dirpath=os.path.join(setup_module.build_dir, "hardware/simulation/src")
+for file in os.listdir(dirpath):
+    if file.startswith("test") and file.endswith(".expected"):
+        os.remove(os.path.join(dirpath,file))
+
+# Find out correct test.expected filename
+test_file_name='test'
+#Check if setup with INIT_MEM (check if macro exists)
+if next((i for i in setup_module.confs if i['name']=='INIT_MEM'), False) and setup_module.confs['INIT_MEM']['val'] != 'NA':
```
Comment (arturum1) [inline hardware/simulation/sim_setup.py:15]:
```suggestion
macro = next((i for i in setup_module.confs if i['name']=='INIT_MEM'), False)
if macro and macro['val'] != 'NA':
```

File: hardware/simulation/sim_setup.py [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    test_file_name+='_initmem'
+#Check if setup with RUN_EXTMEM (check if macro exists)
+if next((i for i in setup_module.confs if i['name']=='RUN_EXTMEM'), False) and setup_module.confs['RUN_EXTMEM']['val'] != 'NA':
```
Comment (arturum1) [inline hardware/simulation/sim_setup.py:18]:
```suggestion
macro = next((i for i in setup_module.confs if i['name']=='RUN_EXTMEM'), False)
if macro and macro['val'] != 'NA':
```

**Review by arturum1 — CHANGES_REQUESTED** [review]

File: hardware/simulation/sim_setup.py (hunk 6 above)
Comment (arturum1) [reply hardware/simulation/sim_setup.py:15]: I had updated these requested changes previously, but seems they were merged before I updated.
(The previous suggested change had the error `confs['INIT_MEM']`)

**Comment by arturum1** [comment]
Ok, seems correct now. The verilator check for `make setup SETUP_ARGS="INIT_MEM=1"` passed.

**Comment by jjts** [comment]
checks failing, many due to small differences in the expected.

**Pull request merged**